File: src/ruby/hangman.rb
puts "=== Hangman ==="
words = File.new("words.txt", "r").read.split("\n")

loop do
  word = words[rand(0..(words.length - 1))]
  correct = ""
  incorrect = ""

  loop do

    show = ""
    for i in 0..(word.length - 1)
      if correct.include? word[i]
        show += word[i]
      else
        show += "_"
      end
    end
    
    if show == word
      puts "\n========\n   WIN\n========\n"
      break
    end
    
    print "\n", show, "\nChances: ", 6 - incorrect.length, "\nCorrect: ", correct, "\nIncorrect: ", incorrect, "\nGuess: "
    guess = gets[0]

    if word.include? guess
      if not correct.include? guess
        correct += guess
      end
    else
      if not incorrect.include? guess
        incorrect += guess
      end
    end

    if incorrect.length >= 6
      puts "\n========\n  LOSS\n  The word was '" + word + "'\n========\n"
      break
    end

  end
end

File: src/javascript/hangman.js
const fs = require("fs");
const readline = require("readline");

console.log("\n=== Hangman ===");
var words = fs.readFileSync("words.txt").toString().split("\r\n");

(async function () {
  while (true) {
    var word = words[Math.floor(Math.random() * words.length)];
    var correct = "";
    var incorrect = "";

    while (true) {
      var show = "";
      for (var i in word) {
        if (correct.includes(word[i])) {
          show += word[i];
        } else {
          show += "_";
        }
      }

      if (show === word) {
        console.log("\n========\n   WIN\n========");
        break;
      }

      console.log(
        "\n" +
          show +
          "\nChances: " +
          (6 - incorrect.length) +
          "\nCorrect: " +
          correct +
          "\nIncorrect: " +
          incorrect,
      );

      var guess = await new Promise(resolve => {
        var rl = readline.createInterface(process.stdin, process.stdout);
        rl.question("Guess: ", res => {
          resolve(res);
          rl.close();
        });
      });

      if (word.includes(guess)) {
        if (!correct.includes(guess)) {
          correct += guess;
        }
      } else {
        if (!incorrect.includes(guess)) {
          incorrect += guess;
        }
      }

      if (incorrect.length >= 6) {
        console.log(`\n========\n  LOSS\n  The word was '${word}'\n========`);
        break;
      }
    }
  }
})();
